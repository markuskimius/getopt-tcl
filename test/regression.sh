#!/bin/sh
#
# Regression test script for getopt-tcl, the getopt library for Tcl.
# https://github.com/markuskimius/getopt-tcl
#
# Copyright Mark K. Kim
# Released under the Apache license 2.0
# https://github.com/markuskimius/getopt-tcl/blob/master/LICENSE
#
#\
exec tclsh "$0" "$@"

lappend auto_path "../lib"
package require getopt


proc main { argv } {
    puts "*** BASIC ***"
    test myarg1
    test -n
    test --no-arg
    test -w warg1
    test --with-arg warg1
    test --with-arg=warg1
    test -i1024
    test -i 1024
    test --integer 1024
    test --integer=1024
    test --opt-arg 128
    test --opt-arg=128

    puts "*** REPETITIONS ***"
    test myarg1 myarg2
    test -nn
    test --no-arg --no-arg
    test -w warg1 -w warg2
    test -wwarg1 -wwarg2
    test --with-arg warg1 --with-arg warg2
    test --with-arg=warg1 --with-arg=warg2
    test -i1024 -i2048
    test -i 1024 -i 2048
    test --integer 1024 --integer 2048
    test --integer=1024 --integer=2048
    test --opt-arg 128 --opt-arg 256
    test --opt-arg=128 --opt-arg=256

    puts "*** COMBINATION (RELATED) ***"
    test -n --no-arg
    test -wwarg1 -w warg2 --with-arg warg3 --with-arg=warg4
    test -i1024 -i 2048 --integer 3072 --integer=4096

    puts "*** COMBINATION (COMPREHENSIVE) ***"
    test -n --no-arg -wwarg1 -w warg2 --with-arg warg3 --with-arg=warg4 -i1024 -i 2048 --integer 3072 --integer=4096 --opt-arg 128 --opt-arg=256 myarg1
    test --no-arg -nwwarg1 -w warg2 --with-arg warg3 --with-arg=warg4 -i1024 -i 2048 --integer 3072 --integer=4096 --opt-arg 128 --opt-arg=256 myarg1
    test --no-arg -wwarg1 -nw warg2 --with-arg warg3 --with-arg=warg4 -i1024 -i 2048 --integer 3072 --integer=4096 --opt-arg 128 --opt-arg=256 myarg1
    test --no-arg -wwarg1 -w warg2 --with-arg warg3 --with-arg=warg4 -ni1024 -i 2048 --integer 3072 --integer=4096 --opt-arg 128 --opt-arg=256 myarg1
    test --no-arg -wwarg1 -w warg2 --with-arg warg3 --with-arg=warg4 -i1024 -ni 2048 --integer 3072 --integer=4096 --opt-arg 128 --opt-arg=256 myarg1
    test --no-arg -wwarg1 -w warg2 --with-arg warg3 --with-arg=warg4 -i1024 -i 2048 --integer 3072 --integer=4096 --opt-arg 128 --opt-arg=256 myarg1 -n
    test --no-arg -wwarg1 -w warg2 --with-arg warg3 --with-arg=warg4 -i1024 -i 2048 --integer 3072 --integer=4096 --opt-arg 128 --opt-arg=256 -n myarg1
    test --no-arg -w warg1 --with-arg warg2 --with-arg=warg3 -i1024 -i 2048 --integer 3072 --integer=4096 --opt-arg 128 --opt-arg=256 myarg1 -nwwarg4
    test --no-arg -w warg1 --with-arg warg2 --with-arg=warg3 -i1024 -i 2048 --integer 3072 --integer=4096 --opt-arg 128 --opt-arg=256 -nwwarg4 myarg1
    test --no-arg -wwarg1 --with-arg warg2 --with-arg=warg3 -i1024 -i 2048 --integer 3072 --integer=4096 --opt-arg 128 --opt-arg=256 myarg1 -nw warg4
    test --no-arg -wwarg1 --with-arg warg2 --with-arg=warg3 -i1024 -i 2048 --integer 3072 --integer=4096 --opt-arg 128 --opt-arg=256 -nw warg4 myarg1
    test --no-arg -wwarg1 -w warg2 --with-arg warg3 --with-arg=warg4 -i 1024 --integer 2048 --integer=3072 --opt-arg 128 --opt-arg=256 myarg1 -ni4096
    test --no-arg -wwarg1 -w warg2 --with-arg warg3 --with-arg=warg4 -i 1024 --integer 2048 --integer=3072 --opt-arg 128 --opt-arg=256 -ni4096 myarg1
    test --no-arg -wwarg1 -w warg2 --with-arg warg3 --with-arg=warg4 -i1024 --integer 2048 --integer=3072 --opt-arg 128 --opt-arg=256 myarg1 -ni 4096
    test --no-arg -wwarg1 -w warg2 --with-arg warg3 --with-arg=warg4 -i1024 --integer 2048 --integer=3072 --opt-arg 128 --opt-arg=256 -ni 4096 myarg1

    puts "*** EMPTY ARGS ***"
    test -n --no-arg -wwarg1 -w "" --with-arg warg2 --with-arg=warg3 -i1024 -i 2048 --integer 3072 --integer=4096 myarg1
    test -n --no-arg -wwarg1 -w warg2 --with-arg "" --with-arg=warg3 -i1024 -i 2048 --integer 3072 --integer=4096 myarg1
    test -n --no-arg -wwarg1 -w warg2 --with-arg warg3 --with-arg= -i1024 -i 2048 --integer 3072 --integer=4096 myarg1
    test -n --no-arg -wwarg1 -w warg2 --with-arg warg3 -i1024 -i 1024 --integer 2048 --integer=3072 myarg1 --with-arg=
    test -n --no-arg -wwarg1 -w warg2 --with-arg warg3 -i1024 -i 1024 --integer 2048 --integer=3072 --with-arg= myarg1
    test -n --no-arg -wwarg1 -w "" --opt-arg warg2 --opt-arg=warg3 -i1024 -i 2048 --integer 3072 --integer=4096 myarg1
    test -n --no-arg -wwarg1 -w warg2 --opt-arg "" --opt-arg=warg3 -i1024 -i 2048 --integer 3072 --integer=4096 myarg1
    test -n --no-arg -wwarg1 -w warg2 --opt-arg warg3 --opt-arg= -i1024 -i 2048 --integer 3072 --integer=4096 myarg1
    test -n --no-arg -wwarg1 -w warg2 --opt-arg warg3 -i1024 -i 1024 --integer 2048 --integer=3072 myarg1 --opt-arg=
    test -n --no-arg -wwarg1 -w warg2 --opt-arg warg3 -i1024 -i 1024 --integer 2048 --integer=3072 --opt-arg= myarg1

    puts "*** EXCEPTIONS (EMPTY INTEGER ARGS) ***"
    test -n --no-arg -wwarg1 -w warg2 --with-arg warg3 --with-arg=warg4 -i1024 -i "" --integer 2048 --integer=3072 myarg1
    test -n --no-arg -wwarg1 -w warg2 --with-arg warg3 --with-arg=warg4 -i1024 -i 2048 --integer "" --integer=3072 myarg1
    test -n --no-arg -wwarg1 -w warg2 --with-arg warg3 --with-arg=warg4 -i1024 -i 2048 --integer 3072 --integer= myarg1

    puts "*** EXCEPTIONS (MISSING MANDATORY ARGS) ***"
    test -n --no-arg -w warg1 --with-arg warg2 --with-arg=warg3 -i1024 -i 1024 --integer 2048 --integer=3072 myarg1 -w
    test -n --no-arg -wwarg1 --with-arg warg2 --with-arg=warg3 -i1024 -i 1024 --integer 2048 --integer=3072 myarg1 -w
    test -n --no-arg -wwarg1 -w warg2 --with-arg=warg3 -i1024 -i 1024 --integer 2048 --integer=3072 myarg1 --with-arg
    test -n --no-arg -wwarg1 -w warg2 --with-arg warg3 --with-arg=warg4 -i 1024 --integer 2048 --integer=3072 myarg1 -i
    test -n --no-arg -wwarg1 -w warg2 --with-arg warg3 --with-arg=warg4 -i1024 --integer 2048 --integer=3072 myarg1 -i
    test -n --no-arg -wwarg1 -w warg2 --with-arg warg3 --with-arg=warg4 -i1024 -i 2048 --integer=3072 myarg1 --integer
    test -n --no-arg -wwarg1 -w warg2 --with-arg warg3 --with-arg=warg4 -i1024 -i 2048 --integer 3072 --integer= myarg1
    test -n --no-arg -wwarg1 -w warg2 --with-arg warg3 --with-arg=warg4 -i1024 -i 2048 --integer 3072 myarg1 --integer=
}


proc test { args } {
    puts "  target $args:"
    array set output [eval target $args]

    foreach name [lsort -dictionary [array names output]] {
        puts "  $name = $output($name)"
    }

    puts ""
}


proc target { args } {
    array set opts [list]

    while { 1 } {
        set c [getopt $args "n 0 no-arg 0 w 1 with-arg 1 i is_int integer is_int opt-arg (is_int,null)"]
        if { $c == -1 } break

        lappend opts(-|opts)  $optopt
        lappend opts(-|index) $optind

        switch -exact -- $c {
            -            { lappend opts(0) $optarg }
            n - no-arg   { lappend opts(n) $optarg }
            w - with-arg { lappend opts(w) $optarg }
            i - integer  { lappend opts(i) $optarg }
            o - opt-arg  { lappend opts(o) $optarg }
            default      { lappend opts(x) $optarg }
        }
    }

    return [array get opts]
}


proc is_int { val } {
    return [string is integer -strict $val]
}


main $argv

# vim:ts=4:sts=4:et:ai
