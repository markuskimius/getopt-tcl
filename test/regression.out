*** BASIC ***
  target myarg1:
  -|index = 1
  -|opts = 0
  0 = myarg1

  target -n:
  -|index = 1
  -|opts = n
  n = {}

  target --no-arg:
  -|index = 1
  -|opts = no-arg
  n = {}

  target -w warg1:
  -|index = 2
  -|opts = w
  w = warg1

  target --with-arg warg1:
  -|index = 2
  -|opts = with-arg
  w = warg1

  target --with-arg=warg1:
  -|index = 1
  -|opts = with-arg
  w = warg1

  target -i1024:
  -|index = 1
  -|opts = i
  i = 1024

  target -i 1024:
  -|index = 2
  -|opts = i
  i = 1024

  target --integer 1024:
  -|index = 2
  -|opts = integer
  i = 1024

  target --integer=1024:
  -|index = 1
  -|opts = integer
  i = 1024

  target --opt-arg 128:
  -|index = 1 2
  -|opts = opt-arg 0
  0 = 128
  o = null

  target --opt-arg=128:
  -|index = 1
  -|opts = opt-arg
  o = 128

*** REPETITIONS ***
  target myarg1 myarg2:
  -|index = 1 2
  -|opts = 0 0
  0 = myarg1 myarg2

  target -nn:
  -|index = 0 1
  -|opts = n n
  n = {} {}

  target --no-arg --no-arg:
  -|index = 1 2
  -|opts = no-arg no-arg
  n = {} {}

  target -w warg1 -w warg2:
  -|index = 2 4
  -|opts = w w
  w = warg1 warg2

  target -wwarg1 -wwarg2:
  -|index = 1 2
  -|opts = w w
  w = warg1 warg2

  target --with-arg warg1 --with-arg warg2:
  -|index = 2 4
  -|opts = with-arg with-arg
  w = warg1 warg2

  target --with-arg=warg1 --with-arg=warg2:
  -|index = 1 2
  -|opts = with-arg with-arg
  w = warg1 warg2

  target -i1024 -i2048:
  -|index = 1 2
  -|opts = i i
  i = 1024 2048

  target -i 1024 -i 2048:
  -|index = 2 4
  -|opts = i i
  i = 1024 2048

  target --integer 1024 --integer 2048:
  -|index = 2 4
  -|opts = integer integer
  i = 1024 2048

  target --integer=1024 --integer=2048:
  -|index = 1 2
  -|opts = integer integer
  i = 1024 2048

  target --opt-arg 128 --opt-arg 256:
  -|index = 1 2 3 4
  -|opts = opt-arg 0 opt-arg 0
  0 = 128 256
  o = null null

  target --opt-arg=128 --opt-arg=256:
  -|index = 1 2
  -|opts = opt-arg opt-arg
  o = 128 256

*** COMBINATION (RELATED) ***
  target -n --no-arg:
  -|index = 1 2
  -|opts = n no-arg
  n = {} {}

  target -wwarg1 -w warg2 --with-arg warg3 --with-arg=warg4:
  -|index = 1 3 5 6
  -|opts = w w with-arg with-arg
  w = warg1 warg2 warg3 warg4

  target -i1024 -i 2048 --integer 3072 --integer=4096:
  -|index = 1 3 5 6
  -|opts = i i integer integer
  i = 1024 2048 3072 4096

*** COMBINATION (COMPREHENSIVE) ***
  target -n --no-arg -wwarg1 -w warg2 --with-arg warg3 --with-arg=warg4 -i1024 -i 2048 --integer 3072 --integer=4096 --opt-arg 128 --opt-arg=256 myarg1:
  -|index = 1 2 3 5 7 8 9 11 13 14 15 16 17 18
  -|opts = n no-arg w w with-arg with-arg i i integer integer opt-arg 0 opt-arg 0
  0 = 128 myarg1
  i = 1024 2048 3072 4096
  n = {} {}
  o = null 256
  w = warg1 warg2 warg3 warg4

  target --no-arg -nwwarg1 -w warg2 --with-arg warg3 --with-arg=warg4 -i1024 -i 2048 --integer 3072 --integer=4096 --opt-arg 128 --opt-arg=256 myarg1:
  -|index = 1 1 2 4 6 7 8 10 12 13 14 15 16 17
  -|opts = no-arg n w w with-arg with-arg i i integer integer opt-arg 0 opt-arg 0
  0 = 128 myarg1
  i = 1024 2048 3072 4096
  n = {} {}
  o = null 256
  w = warg1 warg2 warg3 warg4

  target --no-arg -wwarg1 -nw warg2 --with-arg warg3 --with-arg=warg4 -i1024 -i 2048 --integer 3072 --integer=4096 --opt-arg 128 --opt-arg=256 myarg1:
  -|index = 1 2 2 4 6 7 8 10 12 13 14 15 16 17
  -|opts = no-arg w n w with-arg with-arg i i integer integer opt-arg 0 opt-arg 0
  0 = 128 myarg1
  i = 1024 2048 3072 4096
  n = {} {}
  o = null 256
  w = warg1 warg2 warg3 warg4

  target --no-arg -wwarg1 -w warg2 --with-arg warg3 --with-arg=warg4 -ni1024 -i 2048 --integer 3072 --integer=4096 --opt-arg 128 --opt-arg=256 myarg1:
  -|index = 1 2 4 6 7 7 8 10 12 13 14 15 16 17
  -|opts = no-arg w w with-arg with-arg n i i integer integer opt-arg 0 opt-arg 0
  0 = 128 myarg1
  i = 1024 2048 3072 4096
  n = {} {}
  o = null 256
  w = warg1 warg2 warg3 warg4

  target --no-arg -wwarg1 -w warg2 --with-arg warg3 --with-arg=warg4 -i1024 -ni 2048 --integer 3072 --integer=4096 --opt-arg 128 --opt-arg=256 myarg1:
  -|index = 1 2 4 6 7 8 8 10 12 13 14 15 16 17
  -|opts = no-arg w w with-arg with-arg i n i integer integer opt-arg 0 opt-arg 0
  0 = 128 myarg1
  i = 1024 2048 3072 4096
  n = {} {}
  o = null 256
  w = warg1 warg2 warg3 warg4

  target --no-arg -wwarg1 -w warg2 --with-arg warg3 --with-arg=warg4 -i1024 -i 2048 --integer 3072 --integer=4096 --opt-arg 128 --opt-arg=256 myarg1 -n:
  -|index = 1 2 4 6 7 8 10 12 13 14 15 16 17 18
  -|opts = no-arg w w with-arg with-arg i i integer integer opt-arg 0 opt-arg 0 n
  0 = 128 myarg1
  i = 1024 2048 3072 4096
  n = {} {}
  o = null 256
  w = warg1 warg2 warg3 warg4

  target --no-arg -wwarg1 -w warg2 --with-arg warg3 --with-arg=warg4 -i1024 -i 2048 --integer 3072 --integer=4096 --opt-arg 128 --opt-arg=256 -n myarg1:
  -|index = 1 2 4 6 7 8 10 12 13 14 15 16 17 18
  -|opts = no-arg w w with-arg with-arg i i integer integer opt-arg 0 opt-arg n 0
  0 = 128 myarg1
  i = 1024 2048 3072 4096
  n = {} {}
  o = null 256
  w = warg1 warg2 warg3 warg4

  target --no-arg -w warg1 --with-arg warg2 --with-arg=warg3 -i1024 -i 2048 --integer 3072 --integer=4096 --opt-arg 128 --opt-arg=256 myarg1 -nwwarg4:
  -|index = 1 3 5 6 7 9 11 12 13 14 15 16 16 17
  -|opts = no-arg w with-arg with-arg i i integer integer opt-arg 0 opt-arg 0 n w
  0 = 128 myarg1
  i = 1024 2048 3072 4096
  n = {} {}
  o = null 256
  w = warg1 warg2 warg3 warg4

  target --no-arg -w warg1 --with-arg warg2 --with-arg=warg3 -i1024 -i 2048 --integer 3072 --integer=4096 --opt-arg 128 --opt-arg=256 -nwwarg4 myarg1:
  -|index = 1 3 5 6 7 9 11 12 13 14 15 15 16 17
  -|opts = no-arg w with-arg with-arg i i integer integer opt-arg 0 opt-arg n w 0
  0 = 128 myarg1
  i = 1024 2048 3072 4096
  n = {} {}
  o = null 256
  w = warg1 warg2 warg3 warg4

  target --no-arg -wwarg1 --with-arg warg2 --with-arg=warg3 -i1024 -i 2048 --integer 3072 --integer=4096 --opt-arg 128 --opt-arg=256 myarg1 -nw warg4:
  -|index = 1 2 4 5 6 8 10 11 12 13 14 15 15 17
  -|opts = no-arg w with-arg with-arg i i integer integer opt-arg 0 opt-arg 0 n w
  0 = 128 myarg1
  i = 1024 2048 3072 4096
  n = {} {}
  o = null 256
  w = warg1 warg2 warg3 warg4

  target --no-arg -wwarg1 --with-arg warg2 --with-arg=warg3 -i1024 -i 2048 --integer 3072 --integer=4096 --opt-arg 128 --opt-arg=256 -nw warg4 myarg1:
  -|index = 1 2 4 5 6 8 10 11 12 13 14 14 16 17
  -|opts = no-arg w with-arg with-arg i i integer integer opt-arg 0 opt-arg n w 0
  0 = 128 myarg1
  i = 1024 2048 3072 4096
  n = {} {}
  o = null 256
  w = warg1 warg2 warg3 warg4

  target --no-arg -wwarg1 -w warg2 --with-arg warg3 --with-arg=warg4 -i 1024 --integer 2048 --integer=3072 --opt-arg 128 --opt-arg=256 myarg1 -ni4096:
  -|index = 1 2 4 6 7 9 11 12 13 14 15 16 16 17
  -|opts = no-arg w w with-arg with-arg i integer integer opt-arg 0 opt-arg 0 n i
  0 = 128 myarg1
  i = 1024 2048 3072 4096
  n = {} {}
  o = null 256
  w = warg1 warg2 warg3 warg4

  target --no-arg -wwarg1 -w warg2 --with-arg warg3 --with-arg=warg4 -i 1024 --integer 2048 --integer=3072 --opt-arg 128 --opt-arg=256 -ni4096 myarg1:
  -|index = 1 2 4 6 7 9 11 12 13 14 15 15 16 17
  -|opts = no-arg w w with-arg with-arg i integer integer opt-arg 0 opt-arg n i 0
  0 = 128 myarg1
  i = 1024 2048 3072 4096
  n = {} {}
  o = null 256
  w = warg1 warg2 warg3 warg4

  target --no-arg -wwarg1 -w warg2 --with-arg warg3 --with-arg=warg4 -i1024 --integer 2048 --integer=3072 --opt-arg 128 --opt-arg=256 myarg1 -ni 4096:
  -|index = 1 2 4 6 7 8 10 11 12 13 14 15 15 17
  -|opts = no-arg w w with-arg with-arg i integer integer opt-arg 0 opt-arg 0 n i
  0 = 128 myarg1
  i = 1024 2048 3072 4096
  n = {} {}
  o = null 256
  w = warg1 warg2 warg3 warg4

  target --no-arg -wwarg1 -w warg2 --with-arg warg3 --with-arg=warg4 -i1024 --integer 2048 --integer=3072 --opt-arg 128 --opt-arg=256 -ni 4096 myarg1:
  -|index = 1 2 4 6 7 8 10 11 12 13 14 14 16 17
  -|opts = no-arg w w with-arg with-arg i integer integer opt-arg 0 opt-arg n i 0
  0 = 128 myarg1
  i = 1024 2048 3072 4096
  n = {} {}
  o = null 256
  w = warg1 warg2 warg3 warg4

*** EMPTY ARGS ***
  target -n --no-arg -wwarg1 -w {} --with-arg warg2 --with-arg=warg3 -i1024 -i 2048 --integer 3072 --integer=4096 myarg1:
  -|index = 1 2 3 5 7 8 9 11 13 14 15
  -|opts = n no-arg w w with-arg with-arg i i integer integer 0
  0 = myarg1
  i = 1024 2048 3072 4096
  n = {} {}
  w = warg1 {} warg2 warg3

  target -n --no-arg -wwarg1 -w warg2 --with-arg {} --with-arg=warg3 -i1024 -i 2048 --integer 3072 --integer=4096 myarg1:
  -|index = 1 2 3 5 7 8 9 11 13 14 15
  -|opts = n no-arg w w with-arg with-arg i i integer integer 0
  0 = myarg1
  i = 1024 2048 3072 4096
  n = {} {}
  w = warg1 warg2 {} warg3

  target -n --no-arg -wwarg1 -w warg2 --with-arg warg3 --with-arg= -i1024 -i 2048 --integer 3072 --integer=4096 myarg1:
  -|index = 1 2 3 5 7 8 9 11 13 14 15
  -|opts = n no-arg w w with-arg with-arg i i integer integer 0
  0 = myarg1
  i = 1024 2048 3072 4096
  n = {} {}
  w = warg1 warg2 warg3 {}

  target -n --no-arg -wwarg1 -w warg2 --with-arg warg3 -i1024 -i 1024 --integer 2048 --integer=3072 myarg1 --with-arg=:
  -|index = 1 2 3 5 7 8 10 12 13 14 15
  -|opts = n no-arg w w with-arg i i integer integer 0 with-arg
  0 = myarg1
  i = 1024 1024 2048 3072
  n = {} {}
  w = warg1 warg2 warg3 {}

  target -n --no-arg -wwarg1 -w warg2 --with-arg warg3 -i1024 -i 1024 --integer 2048 --integer=3072 --with-arg= myarg1:
  -|index = 1 2 3 5 7 8 10 12 13 14 15
  -|opts = n no-arg w w with-arg i i integer integer with-arg 0
  0 = myarg1
  i = 1024 1024 2048 3072
  n = {} {}
  w = warg1 warg2 warg3 {}

  target -n --no-arg -wwarg1 -w {} --opt-arg warg2 --opt-arg=warg3 -i1024 -i 2048 --integer 3072 --integer=4096 myarg1:
./regression.sh: invalid argument to option 'opt-arg' -- 'warg3'
  -|index = 1 2 3 5 6 7 8 9 11 13 14 15
  -|opts = n no-arg w w opt-arg 0 opt-arg i i integer integer 0
  0 = warg2 myarg1
  i = 1024 2048 3072 4096
  n = {} {}
  o = null
  w = warg1 {}
  x = warg3

  target -n --no-arg -wwarg1 -w warg2 --opt-arg {} --opt-arg=warg3 -i1024 -i 2048 --integer 3072 --integer=4096 myarg1:
./regression.sh: invalid argument to option 'opt-arg' -- 'warg3'
  -|index = 1 2 3 5 6 7 8 9 11 13 14 15
  -|opts = n no-arg w w opt-arg 0 opt-arg i i integer integer 0
  0 = {} myarg1
  i = 1024 2048 3072 4096
  n = {} {}
  o = null
  w = warg1 warg2
  x = warg3

  target -n --no-arg -wwarg1 -w warg2 --opt-arg warg3 --opt-arg= -i1024 -i 2048 --integer 3072 --integer=4096 myarg1:
  -|index = 1 2 3 5 6 7 8 9 11 13 14 15
  -|opts = n no-arg w w opt-arg 0 opt-arg i i integer integer 0
  0 = warg3 myarg1
  i = 1024 2048 3072 4096
  n = {} {}
  o = null null
  w = warg1 warg2

  target -n --no-arg -wwarg1 -w warg2 --opt-arg warg3 -i1024 -i 1024 --integer 2048 --integer=3072 myarg1 --opt-arg=:
  -|index = 1 2 3 5 6 7 8 10 12 13 14 15
  -|opts = n no-arg w w opt-arg 0 i i integer integer 0 opt-arg
  0 = warg3 myarg1
  i = 1024 1024 2048 3072
  n = {} {}
  o = null null
  w = warg1 warg2

  target -n --no-arg -wwarg1 -w warg2 --opt-arg warg3 -i1024 -i 1024 --integer 2048 --integer=3072 --opt-arg= myarg1:
  -|index = 1 2 3 5 6 7 8 10 12 13 14 15
  -|opts = n no-arg w w opt-arg 0 i i integer integer opt-arg 0
  0 = warg3 myarg1
  i = 1024 1024 2048 3072
  n = {} {}
  o = null null
  w = warg1 warg2

*** EXCEPTIONS (EMPTY INTEGER ARGS) ***
  target -n --no-arg -wwarg1 -w warg2 --with-arg warg3 --with-arg=warg4 -i1024 -i {} --integer 2048 --integer=3072 myarg1:
./regression.sh: invalid argument to option 'i' -- ''
  -|index = 1 2 3 5 7 8 9 11 13 14 15
  -|opts = n no-arg w w with-arg with-arg i i integer integer 0
  0 = myarg1
  i = 1024 2048 3072
  n = {} {}
  w = warg1 warg2 warg3 warg4
  x = {}

  target -n --no-arg -wwarg1 -w warg2 --with-arg warg3 --with-arg=warg4 -i1024 -i 2048 --integer {} --integer=3072 myarg1:
./regression.sh: invalid argument to option 'integer' -- ''
  -|index = 1 2 3 5 7 8 9 11 13 14 15
  -|opts = n no-arg w w with-arg with-arg i i integer integer 0
  0 = myarg1
  i = 1024 2048 3072
  n = {} {}
  w = warg1 warg2 warg3 warg4
  x = {}

  target -n --no-arg -wwarg1 -w warg2 --with-arg warg3 --with-arg=warg4 -i1024 -i 2048 --integer 3072 --integer= myarg1:
./regression.sh: invalid argument to option 'integer' -- ''
  -|index = 1 2 3 5 7 8 9 11 13 14 15
  -|opts = n no-arg w w with-arg with-arg i i integer integer 0
  0 = myarg1
  i = 1024 2048 3072
  n = {} {}
  w = warg1 warg2 warg3 warg4
  x = {}

*** EXCEPTIONS (MISSING MANDATORY ARGS) ***
  target -n --no-arg -w warg1 --with-arg warg2 --with-arg=warg3 -i1024 -i 1024 --integer 2048 --integer=3072 myarg1 -w:
./regression.sh: option requires an argument -- 'w'
  -|index = 1 2 4 6 7 8 10 12 13 14 15
  -|opts = n no-arg w with-arg with-arg i i integer integer 0 w
  0 = myarg1
  i = 1024 1024 2048 3072
  n = {} {}
  w = warg1 warg2 warg3
  x = {}

  target -n --no-arg -wwarg1 --with-arg warg2 --with-arg=warg3 -i1024 -i 1024 --integer 2048 --integer=3072 myarg1 -w:
./regression.sh: option requires an argument -- 'w'
  -|index = 1 2 3 5 6 7 9 11 12 13 14
  -|opts = n no-arg w with-arg with-arg i i integer integer 0 w
  0 = myarg1
  i = 1024 1024 2048 3072
  n = {} {}
  w = warg1 warg2 warg3
  x = {}

  target -n --no-arg -wwarg1 -w warg2 --with-arg=warg3 -i1024 -i 1024 --integer 2048 --integer=3072 myarg1 --with-arg:
./regression.sh: option requires an argument -- 'with-arg'
  -|index = 1 2 3 5 6 7 9 11 12 13 14
  -|opts = n no-arg w w with-arg i i integer integer 0 with-arg
  0 = myarg1
  i = 1024 1024 2048 3072
  n = {} {}
  w = warg1 warg2 warg3
  x = {}

  target -n --no-arg -wwarg1 -w warg2 --with-arg warg3 --with-arg=warg4 -i 1024 --integer 2048 --integer=3072 myarg1 -i:
./regression.sh: option requires an argument -- 'i'
  -|index = 1 2 3 5 7 8 10 12 13 14 15
  -|opts = n no-arg w w with-arg with-arg i integer integer 0 i
  0 = myarg1
  i = 1024 2048 3072
  n = {} {}
  w = warg1 warg2 warg3 warg4
  x = {}

  target -n --no-arg -wwarg1 -w warg2 --with-arg warg3 --with-arg=warg4 -i1024 --integer 2048 --integer=3072 myarg1 -i:
./regression.sh: option requires an argument -- 'i'
  -|index = 1 2 3 5 7 8 9 11 12 13 14
  -|opts = n no-arg w w with-arg with-arg i integer integer 0 i
  0 = myarg1
  i = 1024 2048 3072
  n = {} {}
  w = warg1 warg2 warg3 warg4
  x = {}

  target -n --no-arg -wwarg1 -w warg2 --with-arg warg3 --with-arg=warg4 -i1024 -i 2048 --integer=3072 myarg1 --integer:
./regression.sh: option requires an argument -- 'integer'
  -|index = 1 2 3 5 7 8 9 11 12 13 14
  -|opts = n no-arg w w with-arg with-arg i i integer 0 integer
  0 = myarg1
  i = 1024 2048 3072
  n = {} {}
  w = warg1 warg2 warg3 warg4
  x = {}

  target -n --no-arg -wwarg1 -w warg2 --with-arg warg3 --with-arg=warg4 -i1024 -i 2048 --integer 3072 --integer= myarg1:
./regression.sh: invalid argument to option 'integer' -- ''
  -|index = 1 2 3 5 7 8 9 11 13 14 15
  -|opts = n no-arg w w with-arg with-arg i i integer integer 0
  0 = myarg1
  i = 1024 2048 3072
  n = {} {}
  w = warg1 warg2 warg3 warg4
  x = {}

  target -n --no-arg -wwarg1 -w warg2 --with-arg warg3 --with-arg=warg4 -i1024 -i 2048 --integer 3072 myarg1 --integer=:
./regression.sh: invalid argument to option 'integer' -- ''
  -|index = 1 2 3 5 7 8 9 11 13 14 15
  -|opts = n no-arg w w with-arg with-arg i i integer 0 integer
  0 = myarg1
  i = 1024 2048 3072
  n = {} {}
  w = warg1 warg2 warg3 warg4
  x = {}

